src/mame/misc/anes.cpp

```diff
@@ -38,150 +39,461 @@ namespace {
+
+	m_bank = data & 0x07;
+
+	// the bank change must happen AFTER the next instruction, be it a jmp or a ret
+	m_bank_delay = 2;
+}
+
+void anes_state::blit_w(offs_t offset, uint8_t data)
+{
+	m_blit[offset] = data;
+	if (offset == 0x0a)
+		do_blit();
+}
+
+void anes_state::blit_rom_w(offs_t offset, uint8_t data)
+{
+	uint8_t which = m_blit[0x0b];
+//  logerror("%s: blit ROM written %02X (%04X <- %02X)\n", machine().describe_context(), which, offset, data);
+	if (which == 1)
+	{
+		m_blit_addr[0] = offset;
+		m_blit_val[0] = data;
+	}
+	else if (which == 2)
+	{
+		m_blit_addr[1] = offset;
+		m_blit_val[1] = data;
+	}
+	else
+		logerror("%s: Unknown blit ROM which %02X\n", machine().describe_context(), which);
+}
+
+uint8_t anes_state::blit_status_r()
+{
+	// bits 0-3 = ?
+	// bit    4 = busy
+	return 0x00;
+}
+
+void anes_state::mux_w(uint8_t data)
 {
-	m_vram_offset[offset] = data;
+	m_mux = data;
+	if (m_mux & (~0x1f))
+		logerror("%s: Unknown mux bit written %02X\n", machine().describe_context(), data);
 }
 
-void anes_state::blit_trigger_w(uint8_t data)
+uint8_t anes_state::key_r(offs_t offset)
 {
-	/*
-	 operation is:
-	 checks for bit 4 to be high in port $16
-	 writes either 0 or 3 to port $04
-	 writes a 0 to port $00
-	 writes an offset to ports $0c / $0d / $0e
-	 writes 1 to port $0b, writes to program space, writes 2 to port $0b, writes to program space
-	 writes a mode to port $0b, writes to trigger port $0a
-	 */
+	for (int i = 0; i < 5; i++)
+		if (!BIT(m_mux, i))
+			return m_key[offset][i]->read();
 
-	//printf("%02x%02x%02x\n",m_vram_offset[0],m_vram_offset[1],m_vram_offset[2]);
+	//  logerror("%s: Unknown key read with mux = %02X\n", machine().describe_context(), m_mux);
+	//  return 0xff;
+	return offset ? m_coin->read() : 0xff;
+}
+
+
+uint8_t anes_state::fixed_rom_r(offs_t offset)
+{
+	uint8_t ret = m_maincpu->space(AS_PROGRAM).read_byte(offset);
+	if (!machine().side_effects_disabled())
+	{
+		if (m_bank_delay)
+		{
+			logerror("reading non-banked rom area with bank delay %04x: %02x\n", offset, ret);
+			m_banktimer->adjust(attotime::from_usec(1));
+		}
+	}
+	return ret;
+}
+
+uint8_t anes_state::banked_rom_r(offs_t offset)
+{
+	uint8_t ret = m_maincpu->space(AS_PROGRAM).read_byte(offset + 0x2000);
+	if (!machine().side_effects_disabled())
+	{
+		if (m_bank_delay)
+		{
+			logerror("reading banked rom area with bank delay %04x: %02x\n", offset + 0x2000, ret);
+			m_banktimer->adjust(attotime::from_usec(1));
+		}
+	}
+	return ret;
+}
+
+
+void anes_state::opcodes_map(address_map &map)
+{
+	map(0x0000, 0x1fff).r(FUNC(anes_state::fixed_rom_r));
+	map(0x2000, 0xefff).r(FUNC(anes_state::banked_rom_r));
 }
 
 void anes_state::prg_map(address_map &map)
 {
-	map(0x0000, 0xefff).rom();
-	map(0xf000, 0xffff).ram();
+	map(0x0000, 0x1fff).rom();
+	map(0x2000, 0xefff).bankr(m_rombank);
+	map(0x0000, 0xefff).w(FUNC(anes_state::blit_rom_w));
+	map(0xf000, 0xf7ff).ram();
+	map(0xf900, 0xffff).ram();
 }
```

Comment:
Can you use a single function covering the entire space for `AS_OPCODES` that reads the actual data from the `AS_PROGRAM` space and updates the bank switch status?  See src/mame/osborne/osborne1.cpp (`osborne1_state::opcode_r`) for an example of a Z80 system that does bank switching on M1 cycles.

Reply:
This was already discussed above. I'm doing exactly that in the solution I gave to Haze.
The only problem is that the debugger shows the wrong opcode, ideally the Z80 emulation would have an M1 devcb.

Reply:
Using separate spaces is more useful to handle systems that have encryption only applied to M1 fetches, etc.  Using a devcb would also cause a significant performance penalty.  At any rate, using a timer isn’t a good approach for something that’s most likely triggered by M1 fetches.